File: Uebung2/Aufgabe_5/SimpleHTTPClient.py
import socket
import chardet

host = 'www.google.de'
port = 80
path = '/index.html'

try:
    # Socket-Verbindung zum angegebenen Host und Port aufbauen
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
        client_socket.connect((host, port))

        # Senden des GET-Anforderungsbefehls an den Server
        httpRequest = f"GET {path} HTTP/1.1\r\n" \
                      f"Host: {host}\r\n" \
                      "Connection: close\r\n\r\n"
        client_socket.sendall(httpRequest.encode())

        # Serverantwort empfangen
        response = b""
        while True:
            data = client_socket.recv(1024)
            if not data:
                break
            response += data

        # Zeichenkodierung der Antwort automatisch erkennen und dekodieren
        detected_encoding = chardet.detect(response)['encoding']
        decoded_response = response.decode(detected_encoding)
        print("Serverantwort:\n")
        print(decoded_response)

except socket.error as e:
    print("Fehler beim Kommunizieren mit dem Server:", e)

File: Uebung2/Aufgabe_5/SimpleHTTPClient.java
package Aufgabe_5;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.Socket;
import java.net.UnknownHostException;

public class SimpleHTTPClient {
    public static void main(String[] args) {
        String host = "www.google.de";
        int port = 80;
        String path = "/index.html";

        try {
            sendHTTPRequest(host, port, path);
        } catch (UnknownHostException e) {
            System.err.println("Unbekannter Host: " + e.getMessage());
        } catch (IOException e) {
            System.err.println("Fehler beim Kommunizieren mit dem Server: " + e.getMessage());
        }
    }

    private static void sendHTTPRequest(String host, int port, String path) throws IOException {
        Socket socket = null;
        OutputStream outputStream = null;
        BufferedReader reader = null;

        try {
            socket = new Socket(host, port);
            outputStream = socket.getOutputStream();

            String httpRequest = "GET " + path + " HTTP/1.1\r\n" +
                    "Host: " + host + "\r\n" +
                    "Connection: close\r\n\r\n";
            outputStream.write(httpRequest.getBytes());
            outputStream.flush();

            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            String line;
            StringBuilder response = new StringBuilder();

            while ((line = reader.readLine()) != null) {
                response.append(line).append("\n");
            }

            System.out.println("Serverantwort:\n" + response.toString());
        } finally {
            if (outputStream != null) {
                outputStream.close();
            }
            if (reader != null) {
                reader.close();
            }
            if (socket != null) {
                socket.close();
            }
        }
    }
}
